File: omron_bp/pdf.py
"""Minimal PDF report writer: summary plus a paginated table (Letter, Helvetica, no dependencies).

Mirrors ``win32/src/pdf.cpp`` and ``report.cpp`` so both tools produce the same document.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from pathlib import Path

from omron_bp.storage import CSV_COLUMNS, TIMESTAMP_FORMAT

PAGE_W, PAGE_H = 612.0, 792.0
MARGIN = 54.0
TITLE_SIZE, BODY_SIZE, TABLE_SIZE, SMALL_SIZE = 16.0, 10.0, 9.5, 8.0
ROW_H, LINE_H = 14.0, 13.0
SYS_HIGH, DIA_HIGH = 140, 90
MAX_ROWS = 100_000
MAX_PAGES = 10_000

# Helvetica / Helvetica-Bold advance widths for WinAnsi 32..126, per 1000 em.
_WIDTHS = [278, 278, 355, 556, 556, 889, 667, 191, 333, 333, 389, 584, 278, 333, 278, 278, 556, 556, 556, 556, 556, 556, 556, 556,
           556, 556, 278, 278, 584, 584, 584, 556, 1015, 667, 667, 722, 722, 667, 611, 778, 722, 278, 500, 667, 556, 833, 722, 778,
           667, 778, 722, 667, 611, 722, 667, 944, 667, 667, 611, 278, 278, 278, 469, 556, 333, 556, 556, 500, 556, 556, 278, 556,
           556, 222, 222, 500, 222, 833, 556, 556, 556, 556, 333, 500, 278, 556, 500, 722, 500, 500, 500, 334, 260, 334, 584]
_BOLD_WIDTHS = [278, 333, 474, 556, 556, 889, 722, 238, 333, 333, 389, 584, 278, 333, 278, 278, 556, 556, 556, 556, 556, 556, 556, 556,
                556, 556, 333, 333, 584, 584, 584, 611, 975, 722, 722, 722, 722, 667, 611, 778, 722, 278, 556, 722, 611, 833, 722, 778,
                667, 778, 722, 667, 611, 722, 667, 944, 667, 667, 611, 333, 278, 333, 584, 556, 333, 556, 611, 556, 611, 556, 333, 611,
                611, 278, 278, 556, 278, 889, 611, 611, 611, 611, 389, 556, 333, 611, 556, 778, 556, 556, 500, 389, 280, 389, 584]


@dataclass(frozen=True)
class Column:
    label: str
    x: float
    right_aligned: bool = False


@dataclass
class Report:
    title: str
    subtitle: list[str]
    summary: list[str]
    columns: list[Column]
    rows: list[list[str]]
    footer: str = ""


def text_width(text: str, size: float, bold: bool = False) -> float:
    assert size > 0
    widths = _BOLD_WIDTHS if bold else _WIDTHS
    encoded = text.encode("cp1252", errors="replace")
    return sum(widths[c - 32] if 32 <= c <= 126 else 556 for c in encoded) * size / 1000.0


def _literal(text: str) -> str:
    out = []
    for c in text.encode("cp1252", errors="replace"):
        if c in (0x28, 0x29, 0x5C):  # ( ) backslash
            out.append("\\")
        out.append(chr(c) if c >= 32 else "?")
    return "(" + "".join(out) + ")"


class _Content:
    def __init__(self) -> None:
        self.ops: list[str] = []

    def text(self, text: str, x: float, y: float, size: float, *, bold: bool = False, gray: float = 0.0) -> None:
        self.ops.append(f"{gray:.2f} g BT /{'F2' if bold else 'F1'} {size:.2f} Tf {x:.2f} {y:.2f} Td {_literal(text)} Tj ET")

    def text_right(self, text: str, right: float, y: float, size: float, *, bold: bool = False, gray: float = 0.0) -> None:
        self.text(text, right - text_width(text, size, bold), y, size, bold=bold, gray=gray)

    def line(self, x1: float, y1: float, x2: float, y2: float, *, width: float, gray: float) -> None:
        self.ops.append(f"{gray:.2f} G {width:.2f} w {x1:.2f} {y1:.2f} m {x2:.2f} {y2:.2f} l S")

    def box(self, x: float, y: float, w: float, h: float, gray: float) -> None:
        self.ops.append(f"{gray:.2f} g {x:.2f} {y:.2f} {w:.2f} {h:.2f} re f")

    def stream(self) -> bytes:
        return ("\n".join(self.ops) + "\n").encode("cp1252", errors="replace")


def _draw_header(c: _Content, r: Report) -> float:
    assert r.title and len(r.summary) <= 8
    y = PAGE_H - MARGIN
    c.text(r.title, MARGIN, y - TITLE_SIZE, TITLE_SIZE, bold=True)
    y -= TITLE_SIZE + 10
    for line in r.subtitle:
        c.text(line, MARGIN, y - BODY_SIZE, BODY_SIZE, gray=0.3)
        y -= LINE_H
    if r.summary:
        y -= 6
        box_h = len(r.summary) * LINE_H + 12
        c.box(MARGIN, y - box_h, PAGE_W - 2 * MARGIN, box_h, 0.95)
        ty = y - 8
        for line in r.summary:
            c.text(line, MARGIN + 8, ty - BODY_SIZE, BODY_SIZE)
            ty -= LINE_H
        y -= box_h + 10
    assert y > MARGIN
    return y


def _draw_table_header(c: _Content, r: Report, y: float) -> float:
    for col in r.columns:
        if col.right_aligned:
            c.text_right(col.label, col.x, y - TABLE_SIZE, TABLE_SIZE, bold=True)
        else:
            c.text(col.label, col.x, y - TABLE_SIZE, TABLE_SIZE, bold=True)
    c.line(MARGIN, y - ROW_H + 1, PAGE_W - MARGIN, y - ROW_H + 1, width=0.75, gray=0.6)
    return y - ROW_H - 2


def _draw_footer(c: _Content, r: Report, page: int, pages: int) -> None:
    c.line(MARGIN, MARGIN - 8, PAGE_W - MARGIN, MARGIN - 8, width=0.5, gray=0.75)
    c.text(r.footer, MARGIN, MARGIN - 20, SMALL_SIZE, gray=0.45)
    c.text_right(f"Page {page} of {pages}", PAGE_W - MARGIN, MARGIN - 20, SMALL_SIZE, gray=0.45)


def _rows_that_fit(y: float) -> int:
    available = y - MARGIN - 4
    return 0 if available <= 0 else int(available // ROW_H)


def _page_count(r: Report) -> int:
    assert len(r.rows) <= MAX_ROWS
    probe = _Content()
    fit = _rows_that_fit(_draw_table_header(probe, r, _draw_header(probe, r)))
    remaining, pages = len(r.rows), 1
    for _ in range(MAX_PAGES):
        if remaining <= fit:
            break
        remaining -= fit
        pages += 1
        fit = _rows_that_fit(_draw_table_header(probe, r, PAGE_H - MARGIN))
        assert fit > 0
    return pages


def write_report(path: Path, r: Report) -> int:
    """Write the report; returns the page count."""
    assert 0 < len(r.columns) <= 12
    assert len(r.rows) <= MAX_ROWS
    assert all(len(row) == len(r.columns) for row in r.rows)
    pages = _page_count(r)
    objects: list[bytes] = []  # object 5 onwards; 1-4 are fixed
    page_ids: list[int] = []
    index = 0
    for page in range(1, pages + 1):
        c = _Content()
        y = _draw_header(c, r) if page == 1 else PAGE_H - MARGIN
        y = _draw_table_header(c, r, y)
        fit = _rows_that_fit(y)
        for n in range(fit):
            if index >= len(r.rows):
                break
            if n % 2 == 1:
                c.box(MARGIN, y - ROW_H + 3, PAGE_W - 2 * MARGIN, ROW_H, 0.965)
            for col, cell in zip(r.columns, r.rows[index], strict=True):
                if col.right_aligned:
                    c.text_right(cell, col.x, y - TABLE_SIZE, TABLE_SIZE)
                else:
                    c.text(cell, col.x, y - TABLE_SIZE, TABLE_SIZE)
            y -= ROW_H
            index += 1
        if not r.rows:
            c.text("No readings in this period.", MARGIN, y - TABLE_SIZE, TABLE_SIZE, gray=0.4)
        _draw_footer(c, r, page, pages)
        stream = c.stream()
        objects.append(b"<< /Length " + str(len(stream)).encode() + b" >>\nstream\n" + stream + b"endstream\n")
        stream_id = 4 + len(objects)
        objects.append(
            f"<< /Type /Page /Parent 2 0 R /MediaBox [0 0 {PAGE_W:.2f} {PAGE_H:.2f}] "
            f"/Resources << /Font << /F1 3 0 R /F2 4 0 R >> >> /Contents {stream_id} 0 R >>\n".encode()
        )
        page_ids.append(4 + len(objects))
    assert index == len(r.rows)

    kids = " ".join(f"{p} 0 R" for p in page_ids)
    fixed = [
        b"<< /Type /Catalog /Pages 2 0 R >>\n",
        f"<< /Type /Pages /Kids [{kids}] /Count {pages} >>\n".encode(),
        b"<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica /Encoding /WinAnsiEncoding >>\n",
        b"<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica-Bold /Encoding /WinAnsiEncoding >>\n",
    ]
    out = bytearray(b"%PDF-1.4\n%\xe2\xe3\xcf\xd3\n")
    offsets = [0]
    for number, body in enumerate(fixed + objects, start=1):
        offsets.append(len(out))
        out += f"{number} 0 obj\n".encode() + body + b"endobj\n"
    xref = len(out)
    out += f"xref\n0 {len(offsets)}\n0000000000 65535 f \n".encode()
    for offset in offsets[1:]:
        out += f"{offset:010d} 00000 n \n".encode()
    out += f"trailer\n<< /Size {len(offsets)} /Root 1 0 R >>\nstartxref\n{xref}\n%%EOF\n".encode()
    path.write_bytes(bytes(out))
    return pages


# --- readings report ------------------------------------------------------------


@dataclass(frozen=True)
class ReportOptions:
    scope: str
    period: str
    generated: str
    show_monitor_column: bool = False


def _summary(rows: list[dict[str, str]]) -> list[str]:
    assert 0 < len(rows) <= MAX_ROWS
    sys_ = [int(r["systolic"]) for r in rows]
    dia = [int(r["diastolic"]) for r in rows]
    pulse = [int(r["pulse"]) for r in rows]
    high = sum(1 for s, d in zip(sys_, dia, strict=True) if s >= SYS_HIGH or d >= DIA_HIGH)
    irregular = sum(int(r["irregular_heartbeat"]) for r in rows)
    movement = sum(int(r["movement"]) for r in rows)
    n = len(rows)
    assert high <= n
    flags = f"{high} of {n} readings at or above 140/90"
    if irregular:
        flags += f"   -   {irregular} flagged irregular heartbeat"
    if movement:
        flags += f"   -   {movement} flagged movement"
    return [
        f"{n} reading{'s' if n != 1 else ''}, {rows[0]['timestamp'][:10]} to {rows[-1]['timestamp'][:10]}",
        f"Average {sum(sys_) // n}/{sum(dia) // n} mmHg   (systolic {min(sys_)}-{max(sys_)}, diastolic {min(dia)}-{max(dia)})   "
        f"pulse {sum(pulse) // n} bpm ({min(pulse)}-{max(pulse)})",
        flags,
    ]


def build_report(rows: list[dict[str, str]], options: ReportOptions) -> Report:
    """``rows`` are CSV dict rows (CSV_COLUMNS) sorted oldest first."""
    assert all(tuple(r.keys()) == CSV_COLUMNS for r in rows)
    if options.show_monitor_column:
        columns = [Column("Date", 54), Column("Time", 124), Column("Monitor", 168), Column("Systolic", 300, True),
                   Column("Diastolic", 360, True), Column("Pulse", 412, True), Column("Notes", 432)]
    else:
        columns = [Column("Date", 54), Column("Time", 130), Column("Systolic", 250, True), Column("Diastolic", 320, True),
                   Column("Pulse", 380, True), Column("Notes", 404)]
    table: list[list[str]] = []
    for r in rows:
        when = datetime.strptime(r["timestamp"], TIMESTAMP_FORMAT)
        notes = ", ".join(n for n, flag in (("irregular heartbeat", r["irregular_heartbeat"]), ("movement", r["movement"])) if flag == "1")
        cells = [when.strftime("%Y-%m-%d"), when.strftime("%H:%M")]
        if options.show_monitor_column:
            cells.append(r["device"] + (f" (user {r['user']})" if r["user"] != "1" else ""))
        cells += [r["systolic"], r["diastolic"], r["pulse"], notes]
        table.append(cells)
    assert len(table) == len(rows)
    return Report(
        title="Blood pressure readings",
        subtitle=[
            f"Monitor: {options.scope}",
            f"Period: {options.period}      Generated: {options.generated}      Times are as set on the monitor.",
        ],
        summary=_summary(rows) if rows else [],
        columns=columns,
        rows=table,
        footer="Exported from OMRON monitor memory by OmronBP. Reference: 140/90 mmHg.",
    )


__all__ = ["Column", "Report", "ReportOptions", "build_report", "text_width", "write_report"]

File: win32/src/pdf.cpp
#include "pdf.h"

#include <windows.h>

#include <cassert>
#include <cstdio>
#include <fstream>
#include <stdexcept>

namespace omron::pdf {

namespace {

constexpr float kPageW = 612, kPageH = 792;  // US Letter
constexpr float kMargin = 54;
constexpr float kTitleSize = 16, kBodySize = 10, kTableSize = 9.5f, kSmallSize = 8;
constexpr float kRowH = 14;
constexpr float kLineH = 13;

// Helvetica / Helvetica-Bold advance widths for WinAnsi 32..126, per 1000 em (Adobe AFM).
constexpr int kWidths[95] = {278, 278, 355, 556, 556, 889, 667, 191, 333, 333, 389, 584, 278, 333, 278, 278, 556, 556, 556, 556, 556, 556, 556, 556,
                             556, 556, 278, 278, 584, 584, 584, 556, 1015, 667, 667, 722, 722, 667, 611, 778, 722, 278, 500, 667, 556, 833, 722, 778,
                             667, 778, 722, 667, 611, 722, 667, 944, 667, 667, 611, 278, 278, 278, 469, 556, 333, 556, 556, 500, 556, 556, 278, 556,
                             556, 222, 222, 500, 222, 833, 556, 556, 556, 556, 333, 500, 278, 556, 500, 722, 500, 500, 500, 334, 260, 334, 584};
constexpr int kBoldWidths[95] = {278, 333, 474, 556, 556, 889, 722, 238, 333, 333, 389, 584, 278, 333, 278, 278, 556, 556, 556, 556, 556, 556, 556, 556,
                                 556, 556, 333, 333, 584, 584, 584, 611, 975, 722, 722, 722, 722, 667, 611, 778, 722, 278, 556, 722, 611, 833, 722, 778,
                                 667, 778, 722, 667, 611, 722, 667, 944, 667, 667, 611, 333, 278, 333, 584, 556, 333, 556, 611, 556, 611, 556, 333, 611,
                                 611, 278, 278, 556, 278, 889, 611, 611, 611, 611, 389, 556, 333, 611, 556, 778, 556, 556, 500, 389, 280, 389, 584};

std::string ToWinAnsi(const std::wstring& text) {
    if (text.empty()) return {};
    const int size = WideCharToMultiByte(1252, 0, text.c_str(), static_cast<int>(text.size()), nullptr, 0, "?", nullptr);
    std::string out(static_cast<size_t>(size), '\0');
    WideCharToMultiByte(1252, 0, text.c_str(), static_cast<int>(text.size()), out.data(), size, "?", nullptr);
    return out;
}

// PDF string literal: escape the delimiters.
std::string Literal(const std::wstring& text) {
    std::string out = "(";
    for (unsigned char c : ToWinAnsi(text)) {
        if (c == '(' || c == ')' || c == '\\') out.push_back('\\');
        if (c < 32) c = '?';
        out.push_back(static_cast<char>(c));
    }
    out.push_back(')');
    return out;
}

std::string Num(float v) {
    char buf[32];
    std::snprintf(buf, sizeof buf, "%.2f", v);
    return buf;
}

class Content {
public:
    void Text(const std::wstring& text, float x, float y, float size, bool bold, float gray = 0.0f) {
        stream_ += Num(gray) + " g BT /" + (bold ? "F2" : "F1") + " " + Num(size) + " Tf " + Num(x) + " " + Num(y) + " Td " + Literal(text) + " Tj ET\n";
    }
    void TextRight(const std::wstring& text, float right, float y, float size, bool bold, float gray = 0.0f) {
        Text(text, right - TextWidth(text, size, bold), y, size, bold, gray);
    }
    void Line(float x1, float y1, float x2, float y2, float width, float gray) {
        stream_ += Num(gray) + " G " + Num(width) + " w " + Num(x1) + " " + Num(y1) + " m " + Num(x2) + " " + Num(y2) + " l S\n";
    }
    void Box(float x, float y, float w, float h, float fill_gray) {
        stream_ += Num(fill_gray) + " g " + Num(x) + " " + Num(y) + " " + Num(w) + " " + Num(h) + " re f\n";
    }
    const std::string& Stream() const { return stream_; }

private:
    std::string stream_;
};

class Writer {
public:
    explicit Writer(const std::filesystem::path& path) : out_(path, std::ios::binary | std::ios::trunc) {
        if (!out_) throw std::runtime_error("cannot write " + path.string());
        Emit("%PDF-1.4\n%\xE2\xE3\xCF\xD3\n");
    }

    // Objects 1-4 are reserved: catalog, pages, regular font, bold font.
    size_t AddPage(const Content& content) {
        const size_t stream_obj = NextObject();
        Emit("<< /Length " + std::to_string(content.Stream().size()) + " >>\nstream\n" + content.Stream() + "endstream\n");
        EndObject();
        const size_t page_obj = NextObject();
        Emit("<< /Type /Page /Parent 2 0 R /MediaBox [0 0 " + Num(kPageW) + " " + Num(kPageH) +
             "] /Resources << /Font << /F1 3 0 R /F2 4 0 R >> >> /Contents " + std::to_string(stream_obj) + " 0 R >>\n");
        EndObject();
        pages_.push_back(page_obj);
        return pages_.size();
    }

    void Finish() {
        // Fixed objects are written last; their numbers were reserved up front.
        offsets_[1] = Tell();
        Emit("1 0 obj\n<< /Type /Catalog /Pages 2 0 R >>\nendobj\n");
        offsets_[2] = Tell();
        std::string kids;
        for (size_t p : pages_) kids += std::to_string(p) + " 0 R ";
        Emit("2 0 obj\n<< /Type /Pages /Kids [" + kids + "] /Count " + std::to_string(pages_.size()) + " >>\nendobj\n");
        offsets_[3] = Tell();
        Emit("3 0 obj\n<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica /Encoding /WinAnsiEncoding >>\nendobj\n");
        offsets_[4] = Tell();
        Emit("4 0 obj\n<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica-Bold /Encoding /WinAnsiEncoding >>\nendobj\n");

        const size_t xref = Tell();
        Emit("xref\n0 " + std::to_string(offsets_.size()) + "\n0000000000 65535 f \n");
        for (size_t i = 1; i < offsets_.size(); ++i) {
            char line[24];
            std::snprintf(line, sizeof line, "%010zu 00000 n \n", offsets_[i]);
            Emit(line);
        }
        Emit("trailer\n<< /Size " + std::to_string(offsets_.size()) + " /Root 1 0 R >>\nstartxref\n" + std::to_string(xref) + "\n%%EOF\n");
        out_.flush();
        if (!out_) throw std::runtime_error("PDF write failed");
    }

private:
    size_t NextObject() {
        const size_t number = offsets_.size();
        offsets_.push_back(Tell());
        Emit(std::to_string(number) + " 0 obj\n");
        return number;
    }
    void EndObject() { Emit("endobj\n"); }
    void Emit(const std::string& text) {
        out_.write(text.data(), static_cast<std::streamsize>(text.size()));
        written_ += text.size();
    }
    size_t Tell() const { return written_; }

    std::ofstream out_;
    size_t written_ = 0;
    std::vector<size_t> offsets_{0, 0, 0, 0, 0};  // index = object number; 0 unused
    std::vector<size_t> pages_;
};

// Draws title/subtitle/summary; returns the y where the table may start.
float DrawHeader(Content& c, const Report& r) {
    float y = kPageH - kMargin;
    c.Text(r.title, kMargin, y - kTitleSize, kTitleSize, true);
    y -= kTitleSize + 10;
    for (const std::wstring& line : r.subtitle) {
        c.Text(line, kMargin, y - kBodySize, kBodySize, false, 0.3f);
        y -= kLineH;
    }
    if (!r.summary.empty()) {
        y -= 6;
        const float box_h = static_cast<float>(r.summary.size()) * kLineH + 12;
        c.Box(kMargin, y - box_h, kPageW - 2 * kMargin, box_h, 0.95f);
        float ty = y - 8;
        for (const std::wstring& line : r.summary) {
            c.Text(line, kMargin + 8, ty - kBodySize, kBodySize, false);
            ty -= kLineH;
        }
        y -= box_h + 10;
    }
    return y;
}

float DrawTableHeader(Content& c, const Report& r, float y) {
    for (const Column& col : r.columns) {
        if (col.right_aligned) c.TextRight(col.label, col.x, y - kTableSize, kTableSize, true);
        else c.Text(col.label, col.x, y - kTableSize, kTableSize, true);
    }
    c.Line(kMargin, y - kRowH + 1, kPageW - kMargin, y - kRowH + 1, 0.75f, 0.6f);
    return y - kRowH - 2;
}

void DrawFooter(Content& c, const Report& r, size_t page, size_t pages) {
    c.Line(kMargin, kMargin - 8, kPageW - kMargin, kMargin - 8, 0.5f, 0.75f);
    c.Text(r.footer, kMargin, kMargin - 20, kSmallSize, false, 0.45f);
    c.TextRight(L"Page " + std::to_wstring(page) + L" of " + std::to_wstring(pages), kPageW - kMargin, kMargin - 20, kSmallSize, false, 0.45f);
}

size_t RowsThatFit(float y) {
    const float available = y - kMargin - 4;
    return available <= 0 ? 0 : static_cast<size_t>(available / kRowH);
}

}  // namespace

float TextWidth(const std::wstring& text, float size, bool bold) {
    assert(size > 0);
    const int* widths = bold ? kBoldWidths : kWidths;
    long total = 0;
    for (unsigned char ch : ToWinAnsi(text)) {
        total += (ch >= 32 && ch <= 126) ? widths[ch - 32] : 556;
    }
    return static_cast<float>(total) * size / 1000.0f;
}

size_t WriteReport(const std::filesystem::path& path, const Report& r) {
    assert(!r.columns.empty() && r.columns.size() <= kMaxColumns);
    assert(r.rows.size() <= kMaxRows);
    for (const auto& row : r.rows) {
        if (row.size() != r.columns.size()) throw std::runtime_error("report row has the wrong number of cells");
    }

    // Pass 1: count pages so each footer can say "of N".
    size_t pages = 1;
    {
        Content probe;
        float y = DrawTableHeader(probe, r, DrawHeader(probe, r));
        size_t remaining = r.rows.size();
        size_t fit = RowsThatFit(y);
        for (size_t guard = 0; remaining > fit && guard < kMaxRows; ++guard) {
            remaining -= fit;
            ++pages;
            fit = RowsThatFit(DrawTableHeader(probe, r, kPageH - kMargin));
            assert(fit > 0);
        }
    }

    Writer writer(path);
    size_t index = 0;
    for (size_t page = 1; page <= pages; ++page) {
        Content c;
        float y = page == 1 ? DrawHeader(c, r) : kPageH - kMargin;
        y = DrawTableHeader(c, r, y);
        const size_t fit = RowsThatFit(y);
        for (size_t n = 0; n < fit && index < r.rows.size(); ++n, ++index) {
            if (n % 2 == 1) c.Box(kMargin, y - kRowH + 3, kPageW - 2 * kMargin, kRowH, 0.965f);
            for (size_t col = 0; col < r.columns.size(); ++col) {
                const Column& column = r.columns[col];
                if (column.right_aligned) c.TextRight(r.rows[index][col], column.x, y - kTableSize, kTableSize, false);
                else c.Text(r.rows[index][col], column.x, y - kTableSize, kTableSize, false);
            }
            y -= kRowH;
        }
        if (r.rows.empty()) c.Text(L"No readings in this period.", kMargin, y - kTableSize, kTableSize, false, 0.4f);
        DrawFooter(c, r, page, pages);
        writer.AddPage(c);
    }
    writer.Finish();
    assert(index == r.rows.size());
    return pages;
}

}  // namespace omron::pdf
